cli.sh

```diff
@@ -1251,50 +1280,138 @@ git hooks pull
-        SHARED_HOOKS=$(grep -E "^[^#].+$" <"$(pwd)/.githooks/.shared")
-        update_shared_hooks_in "$SHARED_HOOKS"
+    SHARED_HOOKS=$(git config --global --get githooks.shared)
+    if [ -n "$SHARED_HOOKS" ]; then
+        update_shared_hooks_in --global "$SHARED_HOOKS"
     fi
 
     echo "Finished"
 }
 
 #####################################################
-# Sets the SHARED_ROOT and NORMALIZED_NAME
-#   for the shared hook repo url `$1`.
+# Check if `$1` is not a supported git clone url and
+#   is treated as a local path to a repository.
+#   See `https://tools.ietf.org/html/rfc3986#appendix-B`
+
+# Returns: 0 if it is a local path, 1 otherwise
+#####################################################
+is_local_path() {
+    if echo "$1" | grep -Eq "^[^:/?#]+://" ||  # its a URL `<scheme>://...``
+        echo "$1" | grep -Eq "^.+@.+:.+"; then # or its a short scp syntax
+        return 1
+    fi
+    return 0
+}
+
+#####################################################
+# Check if url `$1`is a local url, e.g `file://`.
+#
+# Returns: 0 if it is a local url, 1 otherwise
+#####################################################
+is_local_url() {
+    if echo "$1" | grep -Eq "^\s*file://"; then
+        return 0
+    fi
+    return 1
+}
+
+#####################################################
+# Sets the `SHARED_ROOT` and `NORMALIZED_NAME`
+#   for the shared hook repo url `$1` and sets
+#   `SHARED_REPO_IS_CLONED` to `true` and its
+#   `SHARED_REPO_CLONE_URL` if is needs to get
+#    cloned and `SHARED_REPO_IS_LOCAL` to `true`
+#    if `$1` points to to a local path.
 #
 # Returns:
-#   0 when successfully finished, 1 otherwise
+#   none
 #####################################################
 set_shared_root() {
-    NORMALIZED_NAME=$(echo "$1" |
-        sed -E "s#.*[:/](.+/.+)\\.git#\\1#" |
-        sed -E "s/[^a-zA-Z0-9]/_/g")
-    SHARED_ROOT="$INSTALL_DIR/shared/$NORMALIZED_NAME"
+
+    SHARED_REPO_CLONE_URL=""
+    SHARED_REPO_CLONE_BRANCH=""
+    SHARED_REPO_IS_LOCAL="false"
+    SHARED_REPO_IS_CLONED="true"
+    DO_SPLIT="true"
+
+    if is_local_path "$1"; then
+        SHARED_REPO_IS_LOCAL="true"
+
+        if is_bare_repo "$1"; then
+            DO_SPLIT="false"
+        else
+            # We have a local path to a non-bare repo
+            SHARED_REPO_IS_CLONED="false"
+            SHARED_ROOT="$1"
+            return
+        fi
+    elif is_local_url "$1"; then
+        SHARED_REPO_IS_LOCAL="true"
+    fi
+
+    # Here we have now supported Git URL or
+    # a local bare-repo `<localpath>`
+
+    # Split "...@(.*)"
+    if [ "$DO_SPLIT" = "true" ] && echo "$1" | grep -q "@"; then
+        SHARED_REPO_CLONE_URL="$(echo "$1" | sed -E "s|^(.*)@.*$|\\1|")"
+        SHARED_REPO_CLONE_BRANCH="$(echo "$1" | sed -E "s|^.*@(.*)$|\\1|")"
+    else
+        SHARED_REPO_CLONE_URL="$1"
+        SHARED_REPO_CLONE_BRANCH=""
+    fi
+
+    # Define the shared clone folder
+    CHECKSUM=$(echo "$1" | md5sum | awk "{ print \$1 }")
```

Comment:
Same as before, make sure that whatever we do for safe checkout name generation, it only uses POSIX.

Reply:
I now use `CHECKSUM=$(echo "$1" | git hash-object --stdin)`

Reply:
Should we maybe use this already in md5sum for the checksums in the hooks.

Reply:
That might be more robust. :-)